pubsub/kubemq/kubemq_events.go

```diff
@@ -0,0 +1,169 @@
+package kubemq
+
+import (
+	"context"
+	"github.com/dapr/components-contrib/pubsub"
+	"github.com/dapr/kit/logger"
+	"github.com/kubemq-io/kubemq-go"
+	"github.com/kubemq-io/kubemq-go/pkg/uuid"
+	"time"
+)
+
+type kubemqEventsClient interface {
+	Stream(ctx context.Context, onError func(err error)) (func(msg *kubemq.Event) error, error)
+	Subscribe(ctx context.Context, request *kubemq.EventsSubscription, onEvent func(msg *kubemq.Event, err error)) error
+	Close() error
+}
+
+type kubeMQEvents struct {
+	client               kubemqEventsClient
+	metadata             *metadata
+	logger               logger.Logger
+	publishFunc          func(event *kubemq.Event) error
+	resultChan           chan error
+	waitForResultTimeout time.Duration
+	ctx                  context.Context
+	ctxCancel            context.CancelFunc
+	isInitialized        bool
+}
+
+func newkubeMQEvents(logger logger.Logger) *kubeMQEvents {
+	return &kubeMQEvents{
+		client:               nil,
+		metadata:             nil,
+		logger:               logger,
+		publishFunc:          nil,
+		resultChan:           make(chan error, 1),
+		waitForResultTimeout: 60 * time.Second,
+		ctx:                  nil,
+		ctxCancel:            nil,
+	}
+}
+func (k *kubeMQEvents) init() error {
+	k.ctx, k.ctxCancel = context.WithCancel(context.Background())
+	if k.metadata.useMock {
+		k.isInitialized = true
+		return nil
+	}
+	clientID := k.metadata.clientID
+	if clientID == "" {
+		clientID = uuid.New()
+	}
+	client, err := kubemq.NewEventsClient(k.ctx,
+		kubemq.WithAddress(k.metadata.host, k.metadata.port),
+		kubemq.WithClientId(clientID),
+		kubemq.WithTransportType(kubemq.TransportTypeGRPC),
+		kubemq.WithCheckConnection(true),
+		kubemq.WithAuthToken(k.metadata.authToken),
+		kubemq.WithAutoReconnect(true),
+		kubemq.WithReconnectInterval(time.Second))
+	if err != nil {
+		k.logger.Errorf("error init kubemq client error: %s", err.Error())
+		return err
+	}
+	k.ctx, k.ctxCancel = context.WithCancel(context.Background())
+	k.client = client
+	if err := k.setPublishStream(); err != nil {
+		k.logger.Errorf("error init kubemq client error: %w", err.Error())
+		return err
+	}
+	k.isInitialized = true
+	return nil
+}
+func (k *kubeMQEvents) Init(meta *metadata) error {
+	k.metadata = meta
+	_ = k.init()
+	return nil
+}
+func (k *kubeMQEvents) setPublishStream() error {
+	var err error
+	k.publishFunc, err = k.client.Stream(k.ctx, func(err error) {
+		select {
+		case k.resultChan <- err:
+		default:
+		}
+	})
+	return err
+}
+func (k *kubeMQEvents) Publish(req *pubsub.PublishRequest) error {
+	if !k.isInitialized {
```

Comment:
This can lead to race conditions if multiple goroutines are calling `Publish` at the same time. Consider using a lock or something else to ensure this is safe for concurrent use.

Reply:
Changed as requested

Reply:
Although this fixes the condition for reading/writing `k.isInitialized`, it doesn't fix the race condition when 2 goroutines are trying to initialize at the same time.

Imagine two goroutines, A and B, calling `Publish` (or `Subscribe`) at the exact same time.

- A loads `k.isInitialized` (atomically) and sees it's false, so calls `init()`
- B loads `k.isInitialized` (atomically) and sees it's false, so calls `init()`
- In both cases, the `init` function is executed till the end before it sets the boolean `k.isInitialized`, so you end up with race conditions there

I think you need to use locks here, so when A starts, it checks if there's a client, then gets a lock, and if that fails, it creates it. At the same time, B is trying to acquire the lock and would block until A is done.

Example is here: https://github.com/dapr/components-contrib/blob/ed745763a0e029f7d7294d867bada99bbcd09827/internal/component/azure/servicebus/client.go#L100-L126

Note that this uses RWMutex locks for performance, which also means that after switching from a read lock to a write loick, it needs to check if the client was created again.

Reply:
Changed to RWMutex as requested